src/Catel.Core/Catel.Core.Shared/EventHandlerExtensions.cs

```diff
@@ -329,8 +329,61 @@ public static bool SafeInvoke(this NotifyCollectionChangedEventHandler handler,
         /// is causing issues.
         /// </summary>
         /// <param name="handler">The handler.</param>
-        /// <param name="args">The arguments.</param>
-        private static void SplitInvoke(Delegate handler, params object[] args)
+        /// <param name="sender">The sender.</param>
+        /// <param name="e">The event args.</param>
+        private static void SplitInvoke(NotifyCollectionChangedEventHandler handler, object sender, NotifyCollectionChangedEventArgs e)
+        {
+            var invocationList = handler.GetInvocationList();
+
+            for (int i = 0; i < invocationList.Length; i++)
+            {
+                try
+                {
+                    var invocationItem = (NotifyCollectionChangedEventHandler) invocationList[i];
+                    invocationItem(sender, e);
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "Failed to invoke event handler at index '{0}'", i);
+                    throw;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Invokes the registered handlers one by one. This way it is easy to determine which subscription on a specific event handler
+        /// is causing issues.
+        /// </summary>
+        /// <param name="handler">The handler.</param>
+        /// <param name="sender">The sender.</param>
+        /// <param name="e">The event args.</param>
+        private static void SplitInvoke(EventHandler handler, object sender, EventArgs e)
+        {
+            var invocationList = handler.GetInvocationList();
+
+            for (int i = 0; i < invocationList.Length; i++)
+            {
+                try
+                {
+                    var invocationItem = (EventHandler)invocationList[i];
```

Comment:
Not sure if this works if we "accidentally" pass in an EventHandler<TEventArgs> instead. I think there should be a fallback to DynamicInvoke if safe cast fails?

And what if we create 1 method like this:

```
private static void SplitInvoke<THandler>(Delegate[] invocationList, Action<THandler> handler) 
{
	 for (int i = 0; i < invocationList.Length; i++)  
	 {  
		 try  
		 {  
			 var invocationItem = invocationList[i] as THandler;  
                         if (invocationItem != null)
                         {
                                handler(invocationItem);
                         }
                         else
                         {
                                 //Dynamic invoke here
                          }
		 }  
		 catch (Exception ex)  
		 {  
			 Log.Error(ex, "Failed to invoke event handler at index '{0}'", i);  
			 throw;  
		 }  
	 }  


}
```